### ISSUE TITLE
Spring Boot JPA with multiple entity managers ignores naming_strategy

### ISSUE THREAD

**Issue by eepstein** [spring-projects/spring-boot#3654]
When using a single data source and entity manager in a Spring Boot JPA project, the hibernate settings are picked up from the application properties (e.g., the application.yml file) and applied as expected.

In the case of multiple data source and entity managers, the naming_strategy setting is not applied and thus we get the default naming strategy.  A minor bug with a likely quick fix I imagine.

I've put together a sample project showing the issue:

https://github.com/eepstein/multids-demo

(which was forked from https://github.com/gratiartis/multids-demo)

The relevant lines are in the FooConfig and BarConfig classes.  For example, from com.sctrcd.multidsdemo.integration.config.foo.FooConfig:

                        // This is to work-around a bug in Spring boot, which is not setting the naming strategy when there
                        // are multiple data sources and entity managers.
                .properties(Collections.singletonMap("hibernate.ejb.naming_strategy",
                        jpaProperties.getHibernate().getNamingStrategy()))

If you comment out that line you'll see an assertion in the test will fail, even though the naming strategy is being correctly set in the application-test.yml file in the test resources folder.

**Comment by snicoll** [comment]
Thanks for the sample! 

Well you are creating the `EntityManager` yourself so there is no magic. Spring Boot can intercept that and apply the naming strategy. We have a related issue to create a sample using two database (#3456) so we could use your sample as a base to show you what needs to be done

(and maybe fix a bug or two along the way ;-) )

**Comment by eepstein** [comment]
Curiously all the other attributes are set automagically via the builders.  This is true for both the DataSource and the EntityManagerFactory.  Plus, the value in question is available via the autowired JpaProperties (which is automatically created by the spring-boot packages).  So, I think this is a bug (or missing feature if you prefer).

There is a need to slap an @Primary on one of the DataSource instances or else spring complains.  But otherwise, and except for the naming_strategy, everything gets wired up from the config/properties file with zero intervention.  I'm of the opinion that naming_strategy should be included in that default auto-wiring.

**Comment by snicoll** [comment]
Compare our [LocalContainerEntityManagerFactoryBean](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/orm/jpa/JpaBaseConfiguration.java#L111) and [yours](https://github.com/eepstein/multids-demo/blob/master/src/main/java/com/sctrcd/multidsdemo/integration/config/bar/BarConfig.java#L43). 

I haven't looked deeper than that but it's fairly different IMO. It's not a terminology problem (I don't have a problem calling something a bug when there is one). I just think your configuration is different. Maybe we can also fix that via a template method. I'll have a look.

**Comment by eepstein** [comment]
Hi Stéphane,

So, this becomes a usage question on my end.  How do I configure things so that the [HibernateJpaAutoConfiguration](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/orm/jpa/HibernateJpaAutoConfiguration.java) is configured and used in the BarConfig class ?    In particular, what's the recommended way to get or setup the EntityManagerFactory so that it is configured with all the vendor-specific settings (Hibernate in this case)?   

Thank you.

**Comment by snicoll** [comment]
Hi @eepstein, no worries! I am not sure yet. I'll have a look tomorrow morning. Cheers!

**Comment by eepstein** [comment]
Perhaps XML config is the way to go in this case?

**Comment by snicoll** [comment]
What leads you to such conclusion? I haven't looked at it yet (last minute M3 stuff kept me busy)

**Comment by snicoll** [comment]
Okay so I have a good overview of the problem now. 

I have [create a sample](https://github.com/snicoll-demos/demo-multi-entity-managers) that exposes several steps. You can browse [the initial version](https://github.com/snicoll-demos/demo-multi-entity-managers/tree/c6333894072ea16e50469ff6031b44c23624e3ea) that is similar to what you end up doing.

Like I said, you are creating the entity manager yourself so any customization that is done in the auto-configuration is lost (since you are overriding it). One of these customizations is to auto-detect the `hbm2ddl.auto` mode (i.e. enable `create` if the database is embedded). The other customization is to set the naming strategy to a sensible value if none was defined. If you look at [the Hibernate customizations of `JpaProperties`](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/orm/jpa/JpaProperties.java#L126) there are two fields really and these are the ones that are customized.

I am not sure I understand why we don't apply these in the builder since we have everything that we need. I'll probably change that.

You can [workaround the issue](https://github.com/snicoll-demos/demo-multi-entity-managers/commit/9e8ed95db5922684559a4a743dfce5c8097a4472) by specifying the following in your configuration

```
spring.jpa.properties.hibernate.ejb.naming_strategy=FooBar
```

Now, what I find a bit disturbing about this setup is that the JPA settings are shared between the two persistence units. What if you want different settings? To demonstrate this, I have [completely disabled the auto-configuration and created a two sets of `JpaProperties`](https://github.com/snicoll-demos/demo-multi-entity-managers/commit/b585747d578e5056a27b289a754ac6d9c330685e) (each on their own namespace)

Finally, that project also showcases how you can [generate the relevant meta-data for those new keys](https://github.com/snicoll-demos/demo-multi-entity-managers/commit/48f1d6ec6e4766cedfd8218e3b34203a8e724a78) so that you get content assistance in the IDE. I hope that helps.

TL;DR the builder should apply the customization internally, I see no reason why it doesn't do it but keep in mind that since you are overriding things, it is normal that you lose some of the auto-magic features.

**Comment by eepstein** [comment]
Hi Stéphane,

So some of that lines up with my own observations and conclusions and some is either new or didn't line up with my own explorations and I'm not sure what I should have done instead.  So, in no particular order:

*  I expected the Builder to have all the default settings baked in, and loaded from the properties.  In short I expected it to be more like the DataSourceBuilder.  It actually seems that the EntityManagerFactoryBuilder does have some things baked in and others not.  This latter is what originally led me to open the issue.
* I agree about the single JpaProperties.  In my particular case it works as the values are shared, but in general something like what's in your sample is the way I'd choose to go.
* I set the spring.jpa.properties.hibernate.ejb.naming_strategy and it had no effect unless applied manually to the builder.  As mentioned this was the originally reason for opening the issue.  Also, although the property name on the hibernate side is hibernate.ejb.naming_strategy, the common way to set it via spring is just:   spring.jpa.hibernate.naming_strategy  (no "properties" and no "ejb" in the name).  It (sort of) matters, since the common way to configure things when there is a single data source and single entity manager is with the shorter name ... and, again, would be nice if the same worked when there are multiple DSs & EMs 
* What I had before is working well now.  In fact changed it to apply all the hibernate properties and changed also to use the regular Spring Boot classes for the primary data source and entity manager creation.  Not ideal, but didn't feel like going the copy & paste route.  May follow what you've done instead.

**Comment by snicoll** [comment]
> I set the spring.jpa.properties.hibernate.ejb.naming_strategy and it had no effect unless applied manually to the builder.

Something should be wrong in your project then since [I got it working](https://github.com/snicoll-demos/demo-multi-entity-managers/commit/9e8ed95db5922684559a4a743dfce5c8097a4472). That demonstrates `ddl-auto` but it works for the naming strategy as well via `spring.jpa.properties.hibernate.ejb.naming_strategy=YourNamingStrategy`

> Should I look for an enhanced builder or some template method in a subsequent release?

We're still discussing [my PR](https://github.com/spring-projects/spring-boot/issues/3707) at this point.

**Comment by eepstein** [comment]
I'd tried 

spring.jpa.hibernate.naming_strategy

That's the one that works with a single DS/EM.  Didn't try the one with .properties. in the path/name.

**Comment by snicoll** [comment]
The key is `spring.jpa.properties.hibernate.ejb.naming_strategy` That key of yours does not map to anything, no surprise it didn't work :)

**Comment by eepstein** [comment]
That's a different key than what one uses when configuring a single DS / EM.
For the single case, per the docs, its:  spring.jpa.hibernate.naming_strategy.

**Comment by snicoll** [comment]
I think you misunderstood my point. Yes, the standard property does not work (that's why this issue exists). I am proposing a *workaround* for the time being.

> You can workaround the issue by specifying the following in your configuration
>
> `spring.jpa.properties.hibernate.ejb.naming_strategy=FooBar`

**Comment by eepstein** [comment]
Thanks.  Yes, work-around is in place and being used in the project.

**Comment by snicoll** [comment]
@wilkinsona if we want to refactor the JPA auto-configuration to be independent of Hibernate somehow, maybe we should take into account the Hibernate 4/5 support as certains keys need to be handled differently according to the hibernate version.

See https://github.com/snicoll/spring-boot/commit/161c629e9fbc22ef967412e618d4f443504ac300

From what I can see, `JpaProperties#Hibernate` may be enough (using some `ClassUtils` hackery).

**Comment by ghostbuster91** [comment]
@snicoll Although it was indeed possible for me to configure custom naming strategy for hibernate following your examples I cannot figure out how to provide a custom naming strategy which needs to be created by spring (because of some dependencies. Could you point me somehow the right direction?

Before going with multi datasource it was as simple as defining a bean:
```
    @Bean
    fun customPhysicalNamingStrategy(@Value("\${TARGET_TABLE}") targetTable: String) = object : SpringPhysicalNamingStrategy() {
        override fun toPhysicalTableName(name: Identifier, context: JdbcEnvironment): Identifier {
            return if (name.text == SynchronizationInfo::class.java.simpleName) {
                Identifier.toIdentifier(targetTable)
            } else {
                super.toPhysicalTableName(name, context)
            }
        }
    }
```
But it no longer works.
Specifying just a property for hibernate obviously fails with cannot find default constructor exception.
I tried attaching it to sessionFactory by `sessionFactoryBuilder.setPhysicalNamingStrategy(myCustomStrategy)` but even when my code was invoked it doesn't seem to have a desired effect. Probably this new session factory must be somehow attached to the custom transactionManager but I have no idea how.

**Comment by ghostbuster91** [comment]
I anybody ever run into similar problem here is how i solved it:

```
    @Primary
    @Bean(name = ["synchronization.entityManagerFactory"])
    fun entityManagerFactory(
            builder: EntityManagerFactoryBuilder,
            @Qualifier("synchronization.dataSource") dataSource: DataSource,
            jpaProperties: JpaProperties,
            @Qualifier("synchronization.namingStrategy") namingStrategy: SpringPhysicalNamingStrategy
    ): LocalContainerEntityManagerFactoryBean {
        return builder
                .dataSource(dataSource)
                .packages(SynchronizationDbConfig::class.java.`package`.name)
                .persistenceUnit(SynchronizationDbConfig::class.java.simpleName)
                .properties(jpaProperties.getHibernateProperties(HibernateSettings().physicalNamingStrategy(namingStrategy)))
                .build()
    }
```

**Comment by petenattress** [comment]
Hi - I appreciate this is an old issue which is probably quite niche. However, I've recently encountered it and it was difficult to debug because I wasn't aware that Spring was applying this naming strategy for me, so the mapping errors I was getting were quite confusing. Lesson learned on that count.

For my use case, it would be ideal if the `EntityManagerFactoryBuilder` came with the magic properties already set (including `ddl-auto` being enabled for an embedded database), so I could create my entity managers as closely as possible to how Spring does by default (this was my assumption about the builder's behaviour from the documentation).

I understand why this may not be possible or desirable, but in any case may I ask that you consider updating [the relevant section of the documentation](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#howto-use-two-entity-managers) to make it clear that the properties must be explicitly passed in to the builder? This would have saved me a fair amount of time! Thank you.

**Comment by wilkinsona** [comment]
@petenattress Could you please open a new issue to capture your suggestions? We can then use it to decide if we should update the docs or make a change to how `EntityManagerFactoryBuilder` currently works.
